### PR TITLE
HW 4139: Make channel limits configurable in the System Console

### PR DESCRIPTION
**Pull request by S4KH**
#### Summary
Made channel limits configurable in the System console

#### Ticket Link
https://github.com/mattermost/platform/issues/4139

#### Checklist
- [x] Changed model/config.go
- [x] Has UI changes
- [x] Includes text changes and localization files updated

### CODE REVIEW

File: config/config.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
         "EnablePostUsernameOverride": false,
         "EnablePostIconOverride": false,
         "EnableTesting": false,
-        "EnableDeveloper": false,
+        "EnableDeveloper": true,
```
Comment (DavidLu1997) [inline config/config.json]: Definitely not intentional on your part, but we should not change `config.json` except to add new fields or change existing defaults.

File: i18n/de.json [hunk 2]
```diff
@@ -3084,6 +3084,10 @@
     "translation": "Ungültige maximale Anzahl an Benutzers pro Team in Teameinstellungen.  Muss eine positive Zahl sein."
   },
   {
+    "id": "model.config.is_valid.max_channels.app_error",
```
Comment (DavidLu1997) [inline i18n/de.json]: Just putting it in `en.json` is enough, our translation server takes care of that

File: model/config.go [hunk 3]
```diff
@@ -223,6 +223,7 @@ type TeamSettings struct {
 	RestrictPublicChannelManagement  *string
 	RestrictPrivateChannelManagement *string
 	UserStatusAwayTimeout            *int64
+	MaxChannelsPerTeam               int
```
Comment (DavidLu1997) [inline model/config.go]: Setting should be a pointer

File: webapp/components/admin_console/users_and_teams_settings.jsx [hunk 4]
```diff
@@ -114,6 +116,24 @@ export default class UsersAndTeamsSettings extends AdminSettings {
                     onChange={this.handleChange}
                 />
                 <TextSetting
+                    id='maxChannelsPerTeam'
```
Comment (DavidLu1997) [inline webapp/components/admin_console/users_and_teams_settings.jsx]: Looks good 👍

File: model/config.go (hunk 3 above)
Comment (S4KH) [inline model/config.go]: I thought it should be similar to what used for MaxUsersPerTeam

File: config/config.json (hunk 1 above)
Comment (S4KH) [inline config/config.json]: I think, I just changed the EnableOpenServer to true from false.

File: model/config.go (hunk 3 above)
Comment (DavidLu1997) [inline model/config.go]: We're trying to migrate to pointers for all config settings, there's a [ticket](https://mattermost.atlassian.net/browse/PLT-3385) for it

File: config/config.json [hunk 5]
```diff
@@ -51,7 +51,8 @@
         "RestrictTeamInvite": "all",
         "RestrictPublicChannelManagement": "all",
         "RestrictPrivateChannelManagement": "all",
-        "UserStatusAwayTimeout": 300
+        "UserStatusAwayTimeout": 300,
+        "MaxChannelsPerTeam": 2111
```
Comment (DavidLu1997) [inline config/config.json]: Set this line to `"MaxChannelsPerTeam": 2000` and remove all the other changes please

**Comment by enahum** [comment]
still have test failing

```
--- FAIL: TestValidTeamName (0.00s)
	team_test.go:98: expect aa as true
	team_test.go:98: expect aaa as true
	team_test.go:98: expect b))b as true
```

**Comment by jasonblais** [comment]
Sorry @S4KH for the delay in testing, there are a few scenarios that need to be tested carefully to make sure everything works as expected.

Planning to continue testing tomorrow.

**Comment by jasonblais** [comment]
Thanks again @S4KH,

I've tested a few cases and I'm not sure if the setting specified in the System Console is enforced. I have specified "2000" as the max channels per team, but I've created more than 2000 channels and would be able to create more.

Would you like to take a second look?

image

image

**Comment by S4KH** [comment]
Got it. Will take a look. I did not touch the functionality on creating teams (it was not mentioned in the steps). Probably have to add checking of maxchannelperteam with current channel number.

**Comment by S4KH** [comment]
Is there a place where we save total number of teams for current channel? I am thinking about adding some validation inside `isTreamCreationAllowed` function. Any suggestions?

**Comment by jasonblais** [comment]
@S4KH I think creating teams is okay, it's that I'm able to create more channels than what is specified in the System Console config setting.

To summarize my note above: For a given team, I was able to create more than the 2000 channels specified [in the config setting](https://github.com/mattermost/platform/pull/4154/files#diff-6a02156604c06becbe5fdca178766accR55). If a limit of 2000 is specified, I should not be able to create more than 2000 channels in a team.

**Comment by enahum** [comment]
just a quick help on this, it appears that all the server side code to prevent the creation of more channels than the one specified in the config is missing

**Comment by S4KH** [comment]
@enahum. Yup, I was actually thinking the same thing. But which file to add that logic?

**Comment by enahum** [comment]
@S4KH it should be here I think
https://github.com/mattermost/platform/blob/master/api/channel.go#L92

**Comment by S4KH** [comment]
Do I also have to add it to the `CreateDirectChannel` function?

**Comment by enahum** [comment]
No, direct channels are for direct messages, there's no limit on those

**Comment by S4KH** [comment]
Got it, thanks :+1:

**Comment by enahum** [comment]
also don't forget to add unit tests in `api/channel_test.go` and if there are any error messages the should have the correspondent string in `i18n/en.json`

**Comment by S4KH** [comment]
Currently, I am trying to use `GetChannelCounts` function inside [sql_channel_store.go](https://github.com/mattermost/platform/blob/master/store/sql_channel_store.go). Is it correct? if so, it returns following

    type StoreResult struct {
	   Data interface{}
	   Err  *model.AppError
    }

    type StoreChannel chan StoreResult

to Data interface this goes in:

    counts := &model.ChannelCounts{Counts: make(map[string]int64), UpdateTimes: make(map[string]int64)}
    for i := range data {
            v := data[i]
	    counts.Counts[v.Id] = v.TotalMsgCount
	    counts.UpdateTimes[v.Id] = v.UpdateAt
    }

    result.Data = counts

So, can you guys suggest a way to get the total number of channels of team  form following code?

`channelCount := <-Srv.Store.Channel().GetChannelCounts(channel.TeamId, channel.CreatorId)`

**Comment by jasonblais** [comment]
Thanks @S4KH for the updates! Looks like there are some build errors, would you like to take a look?

Excited to test it out soon - many have been asking for this new config setting

**Comment by jasonblais** [comment]
Thanks @S4KH! Works great!

Tested on Windows 10, Chrome:

Channel limits are now configurable through **System Console > Users and Teams > Max Channels**. 
 - controls number of public and private channels that can be created per team (i.e. can create at most X number of channels in each team)
 - has no effect on number of direct message channels that can be created
 - if limit is decreased from 2000 to 1 (for instance), there is no impact on existing channels, but users can't create new channels on the team

File: api/channel.go [hunk 6]
```diff
@@ -70,6 +70,21 @@ func createChannel(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		} else {
+			data := result.Data.(*model.ChannelList)
+			if int64(len(data.Channels)+1) > *utils.Cfg.TeamSettings.MaxChannelsPerTeam {
+				c.Err = model.NewLocAppError("createChannel", "api.channel.create_channel.max_channel_limit.app_error", map[string]interface{}{"MaxChannelsPerTeam": *utils.Cfg.TeamSettings.MaxChannelsPerTeam}, "")
+				return
+			}
+		}
+	}
+
 	if channel.Type == model.CHANNEL_OPEN && !HasPermissionToTeamContext(c, channel.TeamId, model.PERMISSION_CREATE_PUBLIC_CHANNEL) {
```
Comment (jwilander) [inline api/channel.go]: Unless I'm missing something, wouldn't it be better to check the permissions before querying the DB? That way if the user doesn't have permission we're not doing any extra work

File: config/config.json (hunk 1 above)
Comment (jwilander) [inline config/config.json]: Can you revert all the changes to config.json except your addition of `MaxChannelsPerTeam`?

**Comment by DavidLu1997** [comment]
You seem to be failing gofmt, a `gofmt -w .` should fix it

**Comment by jwilander** [comment]
@S4KH can you revert any changes in the vendor directory?

**Pull request merged**